lisa/tools/tar.py

```diff
@@ -10,16 +14,60 @@ class Tar(Tool):
     def command(self) -> str:
         return "tar"
 
-    def extract(self, file: str, dest_dir: str) -> str:
+    def extract(
+        self,
+        file: str,
+        dest_dir: str,
+        strip_components: int = 0,
+    ) -> None:
         # create folder when it doesn't exist
+        assert_that(strip_components).described_as(
+            "--strip-components arg for tar must be int >= 0"
+        ).is_greater_than_or_equal_to(0)
         self.node.execute(f"mkdir -p {dest_dir}", shell=True)
-        result = self.run(
-            f"-xvf {file} -C {dest_dir}", shell=True, force_run=True, sudo=True
-        )
+        tar_cmd = f"-xvf {file} -C {dest_dir}"
+        if strip_components:
+            # optionally strip N top level components from a tar file
+            tar_cmd += f" --strip-components={strip_components}"
+        result = self.run(tar_cmd, shell=True, force_run=True, sudo=True)
         if result.exit_code != 0:
             raise LisaException(
                 f"Failed to extract file to {dest_dir}, {result.stderr}"
             )
-        # Get the folder name of the extracted file
-        source_name = result.stdout.split("/")[0]
-        return source_name
+
+    def get_tar_contents(self, file: str) -> List[str]:
+        # return a list of all the files and folders in a tar file
+        result = self.run(
+            f"-tf {file}",
+            shell=True,
+            force_run=True,
+            sudo=True,
+            expected_exit_code=0,
+            expected_exit_code_failure_message=(
+                f"Could not list items in tar file {file}"
+            ),
+        )
+        return result.stdout.splitlines()
+
+    def get_top_level_output_folders(self, file: str) -> List[str]:
+        # tar files can output zero or many output folders
+        # this fetches a list of the directories an extraction will create
+        posix_delimiter = "/"
+        folder_set = set()
+        for tar_contents in self.get_tar_contents(file):
+            if posix_delimiter not in tar_contents:
+                self.node.log.debug(
+                    f"NOTE: tar file {file} contains a regular file "
+                    f"{tar_contents} in the top level output."
+                )
+            else:
+                split_path = tar_contents.split(posix_delimiter)
+                folder_set.add(split_path[0])
+        return list(folder_set)
+
+    def get_single_top_level_output_folder(self, file: str) -> str:
```

Comment:
What's this is used for?

Reply:
For getting the name of the output folder if there is a single one. The old version of Tar.extract assumed there would be a single output folder, since this is pretty common this function would be used to get the output folder name for those cases. The advantage is that this will now assert if they have made a mistake and the tarball has more than one output folder.